File: 0115-distinct-subsequences/0115-distinct-subsequences.py
from collections import defaultdict

class Solution:
  def numDistinct(self, s: str, t: str) -> int:
    tChars = set()
    for c in t:
      tChars.add(c)
    newS = ""
    for c in s:
      if c in tChars:
        newS += c
    if newS == "": return 0
    l, r = 0, len(newS)
    while l < len(newS) and newS[l] != t[0]:
      l += 1
    while r > 0 and newS[r-1] != t[-1]:
      r -= 1
    s = newS[l:r]
    if s == "": return 0
    if s == t: return 1
    
    tally = defaultdict(int)
    for c in s:
      tally[c] += 1
    for c in t:
      if c not in tally: return 0
      if tally[c] == 1:
        tally.pop(c)
      else:
        tally[c] -= 1
        
    dp = [[0] * len(t) for _ in range(len(s))]
    for l in range(len(s)):
      if s[l] == t[0]:
        dp[l][0] += 1
    for r in range(len(t)-1):
      for l in range(r, len(s)):
        if dp[l][r]:
          for i in range(l+1, len(s)):
            if s[i] == t[r+1]:
              dp[i][r+1] += dp[l][r]
    return sum([row[-1] for row in dp])

File: 0115-distinct-subsequences/0115-distinct-subsequences.rb
# @param {String} s
# @param {String} t
# @return {Integer}
def num_distinct(s, t)
  t_chars = Set.new
  t.each_char{ |c| t_chars << c }
  new_s = ""
  s.each_char{ |c| new_s += c if t_chars.include?(c) }
  return 0 if new_s == ""
  l, r = 0, new_s.length
  l += 1 while l < new_s.length && new_s[l] != t[0]
  r -= 1 while r > 0 && new_s[r-1] != t[-1]
  s = new_s[l...r]
  return 0 if s == ""
  return 1 if s == t
    
  tally = Hash.new(0)
  s.each_char{ |c| tally[c] += 1 }
  t.each_char do |c|
    return 0 if !tally.has_key?(c)
    if tally[c] == 1
      tally.delete(c)
    else
      tally[c] -= 1
    end
  end
    
  dp = Array.new(s.length){ [0]*t.length }
  (0...s.length).each{ |l| dp[l][0] += 1 if s[l] == t[0] }
  (0...t.length-1).each do |r|
    (r...s.length).each do |l|
      if dp[l][r] > 0
        (l+1...s.length).each{ |i| dp[i][r+1] += dp[l][r] if s[i] == t[r+1] }
      end
    end
  end
  dp.map{ |row| row[-1] }.sum
end